core/modules/loader/ServerUdpBase.cc

```diff
@@ -0,0 +1,150 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "loader/ServerUdpBase.h"
+
+// System headers
+#include <iostream>
+#include <unistd.h>
+
+// Third-party headers
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.ServerUdpBase");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+std::atomic<uint64_t> ServerUdpBase::_msgIdSeq{1};
+
+
+ServerUdpBase::ServerUdpBase(boost::asio::io_service& io_service, std::string const& host, int port)
+    : _ioService(io_service),
+      _socket(io_service, boost::asio::ip::udp::endpoint(boost::asio::ip::udp::v4(), port)),
+      _hostName(host), _port(port) {
+    _receivePrepare(); // Prime the server for an incoming message.
+}
+
+
+void ServerUdpBase::_receiveCallback(boost::system::error_code const& error, size_t bytesRecvd) {
+    _data->advanceWriteCursor(bytesRecvd); // _data needs to know the valid portion of the buffer.
+    if (!error && bytesRecvd > 0) {
+        LOGS(_log, LOG_LVL_INFO, "rCb received(" << bytesRecvd << "):" <<
+                                 ", code=" << error << ", from endpoint=" << _senderEndpoint);
+
+        _sendData = parseMsg(_data, _senderEndpoint);
+        if (_sendData != nullptr) {
+            _sendResponse();
+        } else {
+            _receivePrepare();
+        }
+    } else {
+        LOGS(_log, LOG_LVL_ERROR, "ServerUdpBase::_receiveCallback got empty message, ignoring");
+        _receivePrepare();
+    }
+
+}
+
+
+void ServerUdpBase::_sendResponse() {
+    _socket.async_send_to(boost::asio::buffer(_sendData->getReadCursor(), _sendData->getBytesLeftToRead()),
+                                              _senderEndpoint,
+                          boost::bind(&ServerUdpBase::_sendCallback, this,
+                                      boost::asio::placeholders::error,
+                                      boost::asio::placeholders::bytes_transferred));
+}
+
+
+void ServerUdpBase::sendBufferTo(std::string const& hostName, int port, BufferUdp& sendBuf) {
+    // The socket is not thread safe. To send on "_socket", it needs to be an async send
+    // and then it needs to know when the message was sent so it can return and free the buffer.
+    using namespace boost::asio;
+
+    std::mutex mtx;
+    std::condition_variable cv;
+    bool done = false;
+
+    auto callbackFunc = [&mtx, &cv, &done](const boost::system::error_code& error, std::size_t bytesTransferred) {
```

Comment:
This is VERY DANGEROUS!!! And this could be a BUG in the application. What you're doing is passing a reference onto a stack variable into a function which is going to be called asynchronously from some other thread at a time when this method will finish and its stack variable will disappear.
You may solve his problem by making these variables data members of the class.

Reply:
In this case it is ok. This function waits for the response before it returns, so all of the references are valid for the life of the callback. I will add a comment.